File: SearchNSort/Squares of a Sorted Array.cpp
class Solution {
public:
    vector<int> sortedSquares(vector<int>& nums) {
        int negPtr = 0;
        int posPtr = nums.size() - 1;
        int origSize = posPtr;
        vector<int> res;
        if (posPtr == negPtr) {
            res.push_back(nums[posPtr] * nums[posPtr]);
            return res;
        }
        while (negPtr <= origSize && posPtr >= 0 && nums[negPtr] < 0 && nums[posPtr] >= 0) {
            int n1 = nums[negPtr] * nums[negPtr];
            int n2 = nums[posPtr] * nums[posPtr];
            if (n1 > n2) {
                res.push_back(n1);
                negPtr++;
            }
            else {
                res.push_back(n2);
                posPtr--;
            }
        }
        while (negPtr <= origSize && nums[negPtr] < 0) {
            res.push_back(nums[negPtr] * nums[negPtr]);
            negPtr++;
        }
        while (posPtr >= 0 && nums[posPtr] >= 0) {
            res.push_back(nums[posPtr] * nums[posPtr]);
            posPtr--;
        }
        reverse(res.begin(), res.end());
        return res;
    }
};

File: Arrays/Squares of a Sorted Array.rs
impl Solution {
    pub fn sorted_squares(nums: Vec<i32>) -> Vec<i32> {
        if nums.len() < 2 {
            return vec!(nums[0].pow(2));
        }

        let mut res = Vec::with_capacity(nums.len());

        let mut startptr:isize = 0;
        let mut endptr:isize = (nums.len() - 1).try_into().unwrap();

        while startptr <= endptr {
            if nums[startptr as usize].abs() >= nums[endptr as usize].abs() {
                res.push(nums[startptr as usize].pow(2));
                startptr += 1;
            }
            else {
                res.push(nums[endptr as usize].pow(2));
                endptr -= 1;
            }
        }

        res.reverse();
        res
    }
}

impl Solution {
    pub fn sorted_squares(nums: Vec<i32>) -> Vec<i32> {
        if nums.len() < 2 {
            return vec!(nums[0].pow(2));
        }

        let mut res = Vec::with_capacity(nums.len());

        let mut midptr = 0;
        for num in nums.iter() {
            if *num >= 0 {break;}
            else {midptr += 1;}
        }
        let mut negptr:isize = midptr-1;
        let mut posptr:isize = midptr;
        while negptr >= 0 && posptr < nums.len() as isize{
            if nums[negptr as usize].abs() > nums[posptr as usize] {
                res.push(nums[posptr as usize].pow(2));
                posptr += 1;
            }
            else {
                res.push(nums[negptr as usize].pow(2));
                negptr -= 1;
            }
        }
        while negptr >= 0 {
            res.push(nums[negptr as usize].pow(2));
            negptr -= 1;
        }
        while posptr < nums.len() as isize { 
            res.push(nums[posptr as usize].pow(2));
            posptr += 1;
        }

        res.reverse();
        res
    }
}
